httpclient5-testing/src/test/java/org/apache/hc/client5/testing/compatibility/sync/CachingHttpClientCompatibilityTest.java

```diff
@@ -0,0 +1,156 @@
+        this.clientResource = new CachingHttpClientResource();
+        this.clientResource.configure(builder -> builder
+                .setCacheConfig(CacheConfig.custom()
+                        .setMaxObjectSize(10240 * 16)
+                        .build())
+                .setResourceFactory(HeapResourceFactory.INSTANCE));
+    }
+
+    CloseableHttpClient client() {
+        return clientResource.client();
+    }
+
+    @Test
+    void test_options_ping() throws Exception {
+        final CloseableHttpClient client = client();
+        final HttpCacheContext context = HttpCacheContext.create();
+        final HttpOptions options = new HttpOptions("*");
+        try (ClassicHttpResponse response = client.executeOpen(target, options, context)) {
+            Assertions.assertEquals(HttpStatus.SC_OK, response.getCode());
+            EntityUtils.consume(response.getEntity());
+        }
+    }
+
+    @Test
+    void test_get_from_cache() throws Exception {
+        final CloseableHttpClient client = client();
+        final String[] resources1 = {"/111", "/222"};
+        for (final String r : resources1) {
+            final HttpCacheContext context1 = HttpCacheContext.create();
+            final HttpGet httpGet1 = new HttpGet(r);
+            try (ClassicHttpResponse response = client.executeOpen(target, httpGet1, context1)) {
+                Assertions.assertEquals(HttpStatus.SC_OK, response.getCode());
+                Assertions.assertEquals(CacheResponseStatus.CACHE_MISS, context1.getCacheResponseStatus());
+                final ResponseCacheControl responseCacheControl = context1.getResponseCacheControl();
+                Assertions.assertNotNull(responseCacheControl);
+                Assertions.assertEquals(600, responseCacheControl.getMaxAge());
+                final HttpCacheEntry cacheEntry = context1.getCacheEntry();
+                Assertions.assertNotNull(cacheEntry);
+                EntityUtils.consume(response.getEntity());
+            }
+            final HttpCacheContext context2 = HttpCacheContext.create();
+            final HttpGet httpGet2 = new HttpGet(r);
+            try (ClassicHttpResponse response = client.executeOpen(target, httpGet2, context2)) {
+                Assertions.assertEquals(HttpStatus.SC_OK, response.getCode());
+                Assertions.assertEquals(CacheResponseStatus.CACHE_HIT, context2.getCacheResponseStatus());
+                final ResponseCacheControl responseCacheControl = context2.getResponseCacheControl();
+                Assertions.assertNotNull(responseCacheControl);
+                Assertions.assertEquals(600, responseCacheControl.getMaxAge());
+                final HttpCacheEntry cacheEntry = context2.getCacheEntry();
+                Assertions.assertNotNull(cacheEntry);
+                Assertions.assertSame(cacheEntry, context1.getCacheEntry());
+                EntityUtils.consume(response.getEntity());
+            }
+
+            Thread.sleep(2000);
```

Comment:
What's this waiting for? Is there something we could check using something like `awaitility` instead of hardcoding a `sleep` which can lead to unpredictable test behaviour.

Reply:
> What's this waiting for? Is there something we could check using something like `awaitility` instead of hardcoding a `sleep` which can lead to unpredictable test behaviour.

@massdosage It is waiting for a cache entry to become stale and trigger its re-validation. I am not sure this can be done without a sleep given this is an integration test. I can however try to reduce it. 

I must admit I do not know what `awaitility` is and whether it can be applied here.

Reply:
@massdosage Have mercy on me.I just discovered `TestContainers` no so long ago and one throws  `awaitility` at me on top of it. 

Seriously our test cases are full of `CountDownLatch` based synchronization code and other ugliness. I would prefer to take a look at  `awaitility` as a way of improving our async testing code overall and not just to replace an occasional `Thread#sleep`

Reply:
Agreed, sounds better to take the approach where you fix more things at once than just this one :) So just leave this "as is" for now. If you want to create an issue summarising where you think something like `awaitility` could improve the tests I might be able to take a look at it.

I'm a big fan of TestContainers, have used it successfully on a number of projects to bring the "Docker Tests" alongside the rest of the Java tests so I'm happy to see it being used here. One just has to be careful to limit the number of these tests as they are slowly and more resource hungry.